include/souper/Infer/AbstractInterpreter.h

```diff
@@ -16,29 +16,75 @@
-    llvm::KnownBits mul(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits udiv(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits urem(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits and_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits or_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits xor_(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits shl(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits lshr(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ashr(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits eq(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ne(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ult(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits slt(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits ule(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
-    llvm::KnownBits sle(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs);
+    llvm::KnownBits add(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits addnsw(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits sub(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits subnsw(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits mul(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits udiv(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits urem(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits and_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits or_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits xor_(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits shl(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits lshr(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ashr(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits eq(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ne(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ult(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits slt(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits ule(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
+    llvm::KnownBits sle(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS);
   }
+
+  bool isConcrete(souper::Inst *I,
+                  bool ConsiderConsts = true,
+                  bool ConsiderHoles = true);
```

Comment:
I think there don't exist any calls with Considerholes = false; if this is true, please eliminate this flag

Reply:
tagging @manasij7479  as it was function signature was added by him originally.

Reply:
I was using it for generating separate experimental results for pruning constants and holes.
We will need that again in future.

Reply:
ugh, but ok